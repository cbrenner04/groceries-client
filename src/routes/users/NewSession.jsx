import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { Button, Form } from 'react-bootstrap';
import axios from 'axios';

import Alert from '../../components/Alert';
import { CheckboxField, EmailField, PasswordField } from '../../components/FormFields';

export default function NewSession(props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [rememberMe, setRememberMe] = useState(false);
  const [errors, setErrors] = useState('');

  useEffect(() => {
    axios.get(`${process.env.REACT_APP_API_BASE}/auth/validate_token`, {
      headers: JSON.parse(sessionStorage.getItem('user')),
    }).then(() => {
      props.history.push('/lists');
    }).catch(() => {
      // noop
    });
  }, [props.history])

  const handleSubmit = event => {
    event.preventDefault();
    setErrors('');
    const user = {
      email,
      password,
      remember_me: rememberMe,
    };
    axios
      .post(`${process.env.REACT_APP_API_BASE}/auth/sign_in`, user)
      .then(({ data: { data }, headers }) => {
        sessionStorage.setItem(
          'user',
          JSON.stringify({
            'access-token': headers['access-token'],
            client: headers['client'],
            uid: data.uid,
          }),
        );
        props.history.push('/');
      })
      .catch(({ response, request, message }) => {
        if (response) {
          setErrors(response.data.errors);
        } else if (request) {
          // TODO: what do here?
        } else {
          setErrors(message);
        }
      });
  };

  return (
    <>
      <Alert errors={errors} handleDismiss={() => setErrors('')} />
      <h2>Log in</h2>
      <Form onSubmit={handleSubmit}>
        <EmailField value={email} handleChange={({ target: { value } }) => setEmail(value)} />
        <PasswordField
          name="password"
          label="Password"
          value={password}
          handleChange={({ target: { value } }) => setPassword(value)}
          placeholder="password"
        />
        <CheckboxField
          name="remember-me"
          classes="mb-3"
          label="Remember me"
          value={rememberMe}
          handleChange={() => setRememberMe(!rememberMe)}
        />
        <Button type="submit" variant="success" block>
          Log in
        </Button>
      </Form>
      <Link to="/users/password/new">Forgot your password?</Link>
    </>
  );
}
