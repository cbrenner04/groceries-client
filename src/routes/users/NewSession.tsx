import React, { type ChangeEvent, type FormEvent, useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Form } from 'react-bootstrap';
import Async, { type PromiseFn } from 'react-async';
import { toast } from 'react-toastify';

import { CheckboxField, EmailField, PasswordField } from 'components/FormFields';
import axios from 'utils/api';
import Loading from 'components/Loading';
import FormSubmission from 'components/FormSubmission';

async function fetchData(fetchParams: { navigate: (url: string) => void }): Promise<void> {
  try {
    await axios.get('/auth/validate_token');
    fetchParams.navigate('/lists');
  } catch {
    // noop. all errors require login
  }
}

export interface INewSessionProps {
  signInUser: (accessToken: string, client: string, uid: string) => void;
}

const NewSession: React.FC<INewSessionProps> = (props): React.JSX.Element => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [rememberMe, setRememberMe] = useState(false);
  const navigate = useNavigate();

  const handleSubmit = async (event: FormEvent<HTMLFormElement>): Promise<void> => {
    event.preventDefault();
    const user = {
      email,
      password,
      remember_me: rememberMe,
    };
    try {
      const response = await axios.post('/auth/sign_in', user);
      props.signInUser(response.headers['access-token'], response.headers.client, response.data.data.uid);
      toast(`Welcome ${email}!`, { type: 'info' });
      navigate('/lists');
    } catch {
      toast('Something went wrong. Please check your credentials and try again.', { type: 'error' });
    }
  };

  return (
    <Async promiseFn={fetchData as unknown as PromiseFn<void>} navigate={navigate}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        <h2>Log in</h2>
        <Form onSubmit={handleSubmit}>
          <EmailField
            value={email}
            handleChange={(event: ChangeEvent<HTMLInputElement>): void => setEmail(event.target.value)}
          />
          <PasswordField
            name="password"
            label="Password"
            value={password}
            handleChange={(event: ChangeEvent<HTMLInputElement>): void => setPassword(event.target.value)}
            placeholder="password"
          />
          <CheckboxField
            name="remember-me"
            classes="mb-3"
            label="Remember me"
            value={rememberMe}
            handleChange={(): void => setRememberMe(!rememberMe)}
          />
          <FormSubmission submitText="Log In" />
        </Form>
        <Link to="/users/password/new">Forgot your password?</Link>
      </Async.Fulfilled>
    </Async>
  );
};

export default NewSession;
