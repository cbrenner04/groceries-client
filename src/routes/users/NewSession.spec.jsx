import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { toast } from 'react-toastify';

import NewSession from './NewSession';
import axios from '../../utils/api';

jest.mock('react-toastify', () => ({
  toast: jest.fn(),
}));

describe('NewSession', () => {
  let props;
  const renderNewSession = (props) => {
    const history = createMemoryHistory();
    return render(
      <Router history={history}>
        <NewSession {...props} />
      </Router>,
    );
  };

  beforeEach(() => {
    props = {
      history: {
        push: jest.fn(),
      },
    };
  });

  it('renders loading when fetch has not completed', () => {
    axios.get = () => new Promise();
    const { container, getByText } = renderNewSession(props);

    expect(container).toMatchSnapshot();
    expect(getByText('Loading...')).toBeVisible();
  });

  it('renders session form when fetch errors', async () => {
    axios.get = jest.fn().mockRejectedValue({ response: { status: 401 } });

    const { container, getByLabelText } = renderNewSession(props);
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(container).toMatchSnapshot();
    expect(getByLabelText('Email')).toBeVisible();
    expect(getByLabelText('Password')).toBeVisible();
    expect(getByLabelText('Remember me')).toBeVisible();
  });

  it('redirects to /lists when fetch is successful', async () => {
    axios.get = jest.fn().mockResolvedValue({});

    renderNewSession(props);
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    expect(props.history.push).toHaveBeenCalledWith('/lists');
  });

  it('creates a new session on successful submission', async () => {
    axios.get = jest.fn().mockRejectedValue({
      response: { status: 401 },
    });
    axios.post = jest.fn().mockResolvedValue({
      data: { data: { uid: 1 } },
      headers: { 'access-token': 'foo', client: 'bar' },
    });
    const spy = jest.spyOn(window.sessionStorage.__proto__, 'setItem');

    const { getByLabelText, getByRole } = renderNewSession(props);
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    fireEvent.change(getByLabelText('Email'), { target: { value: 'foo@example.com' } });
    fireEvent.change(getByLabelText('Password'), { target: { value: 'foo' } });
    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(spy).toHaveBeenCalledWith(
      'user',
      JSON.stringify({
        'access-token': 'foo',
        client: 'bar',
        uid: 1,
      }),
    );
    expect(toast).toHaveBeenCalledWith('Welcome foo@example.com!', { type: 'info' });
    expect(props.history.push).toHaveBeenCalledWith('/lists');
  });

  it('displays error on failed submission', async () => {
    axios.get = jest.fn().mockRejectedValue({ response: { status: 401 } });
    axios.post = jest.fn().mockRejectedValue({ response: { status: 500 } });
    const spy = jest.spyOn(window.sessionStorage.__proto__, 'setItem');

    const { getByLabelText, getByRole } = renderNewSession(props);
    await waitFor(() => expect(axios.get).toHaveBeenCalledTimes(1));

    fireEvent.change(getByLabelText('Email'), { target: { value: 'foo@example.com' } });
    fireEvent.change(getByLabelText('Password'), { target: { value: 'foo' } });
    fireEvent.click(getByLabelText('Remember me'));
    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.post).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('Something went wrong. Please check your credentials and try again.', {
      type: 'error',
    });
    expect(props.history.push).not.toHaveBeenCalled();
    expect(spy).not.toHaveBeenCalled();
  });
});
