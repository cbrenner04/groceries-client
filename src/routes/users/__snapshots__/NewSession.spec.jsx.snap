// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewSession renders loading when fetch has not completed 1`] = `
<div>
  <div
    style="display: flex; justify-content: center; align-items: center;"
  >
    <div
      class="spinner-border"
      role="status"
      style="width: 100px; height: 100px; position: absolute; top: 50%;"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;

exports[`NewSession renders session form when fetch errors 1`] = `
<div>
  <h2>
    Log in
  </h2>
  <form
    class=""
  >
    <div>
      <label
        class="form-label"
        for="email"
      >
        Email
      </label>
      <input
        class="form-control"
        id="email"
        name="email"
        placeholder="jane.smith@example.com"
        type="email"
        value=""
      />
    </div>
    <div>
      <label
        class="form-label"
        for="password"
      >
        Password
      </label>
      <input
        autocomplete="off"
        class="form-control"
        id="password"
        name="password"
        placeholder="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="remember-me"
          name="remember-me"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="remember-me"
          title=""
        >
          Remember me
        </label>
      </div>
    </div>
    <div
      class="d-grid gap-2 mt-3"
    >
      <button
        class="btn btn-success"
        type="submit"
      >
        Log In
      </button>
    </div>
  </form>
  <a
    href="/users/password/new"
  >
    Forgot your password?
  </a>
</div>
`;
